### PR TITLE
Ability to clear list of analysis when no trace is selected (and open)

### LINKED ISSUE #489: Clear list of analysis when no trace is selected (and open)
Leaving the list of analyses visible when a trace tab is closed is confusing - it should be cleared when a tab is closed.

pr486-refresh-analyses

_Originally posted by @mirsky-work in https://github.com/theia-ide/theia-trace-extension/issues/486#issuecomment-923099308_

### PR DESCRIPTION
**Pull request by hriday-panchasara**
fixes #489 

- When no trace tabs are open the available views section is empty:
Screen Shot 2021-11-09 at 10 08 10 AM

Fix:
- Whenever a trace tab is closed fire a signal to trace-explorer-views-widget which checks the number of open tabs. If none, it will clear available views
- Then update available views for the selected experiment only if trace tabs are open

### CODE REVIEW

File: packages/react-components/src/trace-explorer/trace-explorer-views-widget.tsx [hunk 1]
```diff
@@ -42,11 +42,13 @@ export class ReactAvailableViewsWidget extends React.Component<ReactAvailableVie
         });
         signalManager().on(Signals.EXPERIMENT_SELECTED, this.doHandleExperimentSelectedSignal);
         signalManager().on(Signals.CHECK_OPEN_WIDGETS, this.doHandleCheckOpenWidgetsSignal);
-        this.state = { availableOutputDescriptors: [] };
+        this.state = { availableOutputDescriptors: [], lastSelectedOutputIndex: -1 };
```
Comment (PatrickTasse) [inline packages/react-components/src/trace-explorer/trace-explorer-views-widget.tsx:45]: I don't think this commit from master branch should be merged in this branch, instead this branch should be rebased on the latest master branch commit.

File: packages/react-components/src/trace-explorer/trace-explorer-views-widget.tsx [hunk 2]
```diff
@@ -48,7 +48,6 @@ export class ReactAvailableViewsWidget extends React.Component<ReactAvailableVie
     componentWillUnmount(): void {
         signalManager().off(Signals.EXPERIMENT_SELECTED, this.doHandleExperimentSelectedSignal);
         signalManager().off(Signals.CHECK_OPEN_WIDGETS, this.doHandleCheckOpenWidgetsSignal);
-
```
Comment (PatrickTasse) [inline packages/react-components/src/trace-explorer/trace-explorer-views-widget.tsx]: Is this commit a fixup of the first commit? Could they be squashed instead?

**Review by PatrickTasse — CHANGES_REQUESTED** [review]

File: packages/base/src/signals/signal-manager.ts [hunk 3]
```diff
@@ -41,7 +41,8 @@ export const Signals = {
     RESET_ZOOM: 'reset zoom',
     MARKER_CATEGORIES_FETCHED: 'marker categories fetched',
     MARKERSETS_FETCHED: 'markersets fetched',
-    MARKER_CATEGORY_CLOSED: 'marker category closed'
+    MARKER_CATEGORY_CLOSED: 'marker category closed',
+    CHECK_OPEN_WIDGETS: 'check open widgets'
```
Comment (PatrickTasse) [inline packages/base/src/signals/signal-manager.ts:45]: There's too many signals already doing the same thing, I'm a bit confused.

Could you see if you can reuse one of these existing signals:
    EXPERIMENT_CLOSED: 'experiment closed',
    OPENED_TRACES_UPDATED: 'opened traces updated',
    CLOSE_TRACEVIEWERTAB: 'tab closed',

Whichever one is fired in all the cases you care about...

File: packages/react-components/src/trace-explorer/trace-explorer-views-widget.tsx [hunk 4]
```diff
@@ -177,4 +189,3 @@ export class ReactAvailableViewsWidget extends React.Component<ReactAvailableVie
         return outputs;
     }
 }
-
```
Comment (PatrickTasse) [inline packages/react-components/src/trace-explorer/trace-explorer-views-widget.tsx:180]: Revert this change

File: packages/react-components/src/trace-explorer/trace-explorer-views-widget.tsx [hunk 5]
```diff
@@ -188,4 +188,4 @@ export class ReactAvailableViewsWidget extends React.Component<ReactAvailableVie
         outputs = outputs.filter(value => !value.id.includes('scatter'));
         return outputs;
     }
-}
\ No newline at end of file
+}
```
Comment (PatrickTasse) [inline packages/react-components/src/trace-explorer/trace-explorer-views-widget.tsx:191]: Squash this fixup in original commit

**Review by PatrickTasse — CHANGES_REQUESTED** [review]
I don't know if this goes in the scope of this change, but the selected element in Opened Traces widget always follows the trace viewer tab that has focus. So when you close the last tab, the selection in Opened Traces widget should be cleared.

File: packages/base/src/signals/signal-manager.ts (hunk 3 above)
Comment (hriday-panchasara) [reply packages/base/src/signals/signal-manager.ts]: > There's too many signals already doing the same thing, I'm a bit confused.
> 
> Could you see if you can reuse one of these existing signals: EXPERIMENT_CLOSED: 'experiment closed', OPENED_TRACES_UPDATED: 'opened traces updated', CLOSE_TRACEVIEWERTAB: 'tab closed',
> 
> Whichever one is fired in all the cases you care about...

Reworked the code to implement OPENED_TRACES_UPDATED signal

**Comment by hriday-panchasara** [comment]
> I don't know if this goes in the scope of this change, but the selected element in Opened Traces widget always follows the trace viewer tab that has focus. So when you close the last tab, the selection in Opened Traces widget should be cleared.

When all tabs are closed the selection in Opened Traces widget is cleared:
Screen Shot 2021-11-15 at 16 18 52 PM

File: theia-extensions/viewer-prototype/src/browser/trace-viewer/trace-viewer.tsx [hunk 6]
```diff
@@ -250,9 +251,11 @@ export class TraceViewerWidget extends ReactWidget {
         }
     }
 
-    onCloseRequest(msg: Message): void {
+    async onCloseRequest(msg: Message): Promise<void> {
         this.statusBar.removeElement('time-selection-range');
         super.onCloseRequest(msg);
+        const remoteExperiments = await this.experimentManager.getOpenedExperiments();
+        signalManager().fireOpenedTracesChangedSignal(new OpenedTracesUpdatedSignalPayload(remoteExperiments ? remoteExperiments.length : 0));
```
Comment (PatrickTasse) [inline theia-extensions/viewer-prototype/src/browser/trace-viewer/trace-viewer.tsx:258]: Although it works, it's a bit misleading to broadcast this signal here because the opened traces (number of traces in opened traces widget) is not changed by closing a trace viewer tab.

Perhaps it would be better to use CLOSE_TRACEVIEWERTAB signal?

**Review by PatrickTasse — CHANGES_REQUESTED** [review]
There is a merge conflict, please rebase this branch on latest master.

File: theia-extensions/viewer-prototype/src/browser/trace-viewer/trace-viewer.tsx (hunk 6 above)
Comment (hriday-panchasara) [reply theia-extensions/viewer-prototype/src/browser/trace-viewer/trace-viewer.tsx:247]: That was my first guess as well, but I'm not sure how to make it work
- CLOSE_TRACEVIEWERTAB signal is fired by the opened-traces view's context menu and is caught by trace-viewer.tsx to close the tab. 
- The handler for CLOSE_TRACEVIEWERTAB signal in trace-viewer.tsx calls shell.closeWidget to close the tab. But this handler does not catch the event where the user closes the tab using the tab's remove ('x') button
- Widget.onCloseRequest is called whenever a tab is closed, either from the context menu or the tab's remove button. So I wanted to emit an event in this function, but I don't know if I can emit CLOSE_TRACEVIEWERTAB given that trace-viewer.tsx is watching for the same signal.

Comment (PatrickTasse) [reply theia-extensions/viewer-prototype/src/browser/trace-viewer/trace-viewer.tsx:247]: Yes, that was my observation too. I tried emitting the CLOSE_TRACEVIEWERTAB event and it didn't seem to cause any problem. But OK, we can revisit it later if needed. I think there's a few other issues with how signals are used in the application.

File: packages/react-components/src/trace-explorer/trace-explorer-opened-traces-widget.tsx [hunk 7]
```diff
@@ -9,6 +9,7 @@ import { faCopy } from '@fortawesome/free-solid-svg-icons';
 import { OpenedTracesUpdatedSignalPayload } from 'traceviewer-base/lib/signals/opened-traces-updated-signal-payload';
 import { ITspClientProvider } from 'traceviewer-base/lib/tsp-client-provider';
 import { WidgetManager } from '@theia/core/lib/browser';
+import { faTimes } from '@fortawesome/free-solid-svg-icons';
```
Comment (PatrickTasse) [inline packages/react-components/src/trace-explorer/trace-explorer-opened-traces-widget.tsx:12]: Could you please rebase your branch on top of latest master instead of adding a merge commit?

**Review by PatrickTasse — CHANGES_REQUESTED** [review]

File: packages/react-components/src/trace-explorer/trace-explorer-opened-traces-widget.tsx (hunk 7 above)
Comment (hriday-panchasara) [reply packages/react-components/src/trace-explorer/trace-explorer-opened-traces-widget.tsx:11]: > 

Rebased on top of master and fixed merge conflicts. Seems to work since all commits got squashed as well

**Review by PatrickTasse — APPROVED** [review]

**Review by bhufmann — APPROVED** [review]
Thanks ... looks good to me.
